From 2aadf285b2c1628a7dc6b6dee445de41ab2a638d Mon Sep 17 00:00:00 2001
From: Bidipta Samadder <bidiptax.samadder@intel.com>
Date: Wed, 6 May 2015 18:01:47 +0530
Subject: [PATCH] EM: Addtion of NULL character in OEM0 ACPI table batt_id

OEM0 ACPI table - batt_id string was not null terminated.
So one more byte is added to batt_id string of ps_pse_mod_prof
structure. Then added the null character at the end of batt_id
string manually.

Change-Id: I6dc5b9613b009e47298ff463fd3105c70e386181
Tracked-On: https://jira01.devtools.intel.com/browse/AKP-1568
Signed-off-by: Bidipta Samadder <bidiptax.samadder@intel.com>
---
 drivers/platform/x86/intel_em_config.c | 14 +++++++-------
 include/linux/power/battery_id.h       |  5 +++--
 2 files changed, 10 insertions(+), 9 deletions(-)

diff --git a/drivers/platform/x86/intel_em_config.c b/drivers/platform/x86/intel_em_config.c
index 7e65b3e8fab0..2735c25ab9fc 100644
--- a/drivers/platform/x86/intel_em_config.c
+++ b/drivers/platform/x86/intel_em_config.c
@@ -63,18 +63,18 @@ EXPORT_SYMBOL(em_config_get_oem1_data);
 
 static int em_config_get_charge_profile(struct ps_pse_mod_prof *chrg_prof)
 {
+	struct em_config_oem0_data chrg_prof_temp;
 	int ret = 0;
 
 	if (chrg_prof == NULL)
 		return 0;
-	ret = em_config_get_oem0_data((struct em_config_oem0_data *)chrg_prof);
+	ret = em_config_get_oem0_data(&chrg_prof_temp);
 	if (ret > 0) {
-		/*
-		 * battery_type field contains 2 bytes, and upper byte
-		 * contains * battery_type & lower byte used for turbo,
-		 * which is discarded
-		 */
-		chrg_prof->battery_type = chrg_prof->battery_type >> 8;
+		memcpy(chrg_prof, &chrg_prof_temp, BATTID_STR_LEN);
+		chrg_prof->batt_id[BATTID_STR_LEN] = '\0';
+		memcpy(&(chrg_prof->turbo),
+			((char *)&chrg_prof_temp) + BATTID_STR_LEN,
+			sizeof(struct em_config_oem0_data) - BATTID_STR_LEN);
 #ifdef DEBUG
 		dump_chrg_profile(chrg_prof);
 #endif
diff --git a/include/linux/power/battery_id.h b/include/linux/power/battery_id.h
index 50f9993f4eca..592b08457d8a 100644
--- a/include/linux/power/battery_id.h
+++ b/include/linux/power/battery_id.h
@@ -43,9 +43,10 @@ struct ps_temp_chg_table {
 /* Charging Profile */
 struct ps_pse_mod_prof {
 	/* battery id */
-	char batt_id[BATTID_STR_LEN];
+	char batt_id[BATTID_STR_LEN + 1];
+	u8 turbo;
 	/* type of battery */
-	u16 battery_type;
+	u8 battery_type;
 	u16 capacity;
 	u16 voltage_max;
 	/* charge termination current */
-- 
1.9.1

